### PR TITLE
Create moveXXX() functions for MutableList

### PR DESCRIPTION
**Pull request by thomasnield**
A proposal to implement `moveXXX()` operators for `MutableList` to manage the movement of items to different indexes.

### CODE REVIEW

File: proposals/stdlib/mutablelist-move.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+## Summary
+
+It would be helpful to create some additional extension functions for `MutableList` to move items to new indexes, as well as move them up/down, and one or all items that meet a predicate.
+
+These can be implemented using only extension functions on `MutableList`.
+
+## Similar API review
+
+Cannot think of an explicit precendent in another API.
```
Comment (thomasnield) [inline proposals/stdlib/mutablelist-move.md]: I figured there was something out there. Not surprised the `Collections` helper static class has a method, although I've never used it. I will investigate those.

File: proposals/stdlib/mutablelist-move.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+## Alternatives
+
+Developers could implement the extension functions themselves, as shown below. Moving items can be tricky and verbose to coordinate with local implementations.
+
+```kotlin
+/**
+ * Moves the given **T** item to the specified index
+ */
+fun <T> MutableList<T>.move(item: T, newIndex: Int)  {
```
Comment (thomasnield) [inline proposals/stdlib/mutablelist-move.md]: Thought I accounted for that, apparently not!

File: proposals/stdlib/mutablelist-move.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    add(newIndex, item)
+}
+
+/**
+ * Moves the given item at the `oldIndex` to the `newIndex`
+ */
+fun <T> MutableList<T>.move(oldIndex: Int, newIndex: Int)  {
+    val item = this[oldIndex]
+    removeAt(oldIndex)
+    add(newIndex, item)
```
Comment (thomasnield) [inline proposals/stdlib/mutablelist-move.md]: I may need to rename this to `moveAt()` or something similar so the overload doesn't clash if its a `MutableList<Int>`. Is that what you were thinking with previous comment?

Good catch on inconsistent state. It crossed my mind but I reasoned incorrectly as my thought experiments thought of the item only moving up, not down. I may need to do a check to see if an offset occurs and adjust by `1` accordingly.

File: proposals/stdlib/mutablelist-move.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+/**
+ * Moves all items meeting a predicate to the given index
+ */
+fun <T> MutableList<T>.moveAll(newIndex: Int, predicate: (T) -> Boolean) {
+
+    var newIndexIncrement = newIndex
+
+    asSequence().toList().asSequence().withIndex()
```
Comment (thomasnield) [inline proposals/stdlib/mutablelist-move.md]: Yup. That's valid.

File: proposals/stdlib/mutablelist-move.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            .filter { v -> predicate.invoke(v.value) }
+            .forEach { move(it.index, newIndexIncrement++) }
```
Comment (thomasnield) [inline proposals/stdlib/mutablelist-move.md]: My tests didn't see an issue. I think I know what you're saying, and I'm trying to think of what could go wrong here... but I can't think of anything.

File: proposals/stdlib/mutablelist-move.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    add(newIndex, item)
+}
+
+/**
+ * Moves the given element **T** up the **MutableList** by an index increment
+ * unless it is at the bottom already which will result in no movement
+ */
+fun <T> MutableList<T>.moveDown(item: T) {
+    val currentIndex = indexOf(item)
+    val newIndex = (currentIndex + 1)
```
Comment (thomasnield) [inline proposals/stdlib/mutablelist-move.md]: So maybe a `false` needs to be returned to indicate a failure, or even throw an `Exception`.

File: proposals/stdlib/mutablelist-move.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+The proposal is to put this in Standard Library, and make it immediately available wherever `MutableList` is used.
+
+## Reference implementation
+
+A proposed implementation can be [found in this repository](https://github.com/thomasnield/kotlin-collection-extras/blob/master/src/main/kotlin/MutableListMoveExt.kt).
+
+
+## Unresolved questions
+
+* Is there a way to move items without removing and adding them? (Unlikely to my knowledge).
```
Comment (thomasnield) [inline proposals/stdlib/mutablelist-move.md]: Ok.

**Comment by thomasnield** [comment]
I've use some variants of these in an application I'm building, and this is an API I've written from scratch in an attempt to formalize and fill in the gaps. I agree it is not polished. I'll make some changes this weekend and may ask for clarification on a few things, and thanks for taking the time to review this.

File: proposals/stdlib/mutablelist-move.md (hunk 1 above)
Comment (thomasnield) [inline proposals/stdlib/mutablelist-move.md]: It looks like `Collections.rotate()` simply shifts the entire `List` up or down for all elements in their current order. It does not move elements relative to each other.  When I specify a condition, I am moving elements with that condition to a certain area in the `List`. 

`subList()` looks only to extract part of the `List` between two indexes. If there is a way to make it work in conjunction with `Collections.rotate()`, I'm not sure what implementation you are envisioning.

Comment (ilya-g) [inline proposals/stdlib/mutablelist-move.md]: I mean something like this:
```
fun <T> MutableList<T>.move(fromIndex: Int, toIndex: Int) {
    if (toIndex > fromIndex) {
        Collections.rotate(subList(fromIndex, toIndex + 1), -1)    
    } else if (toIndex < fromIndex) {
        Collections.rotate(subList(toIndex, fromIndex + 1), 1)
    }
}
```

File: proposals/stdlib/mutablelist-move.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+fun <T> MutableList<T>.move(oldIndex: Int, newIndex: Int)  {
+    val item = this[oldIndex]
+    removeAt(oldIndex)
+    add(newIndex, item)
+}
+
+/**
+ * Moves all items meeting a predicate to the given index
+ */
+fun <T> MutableList<T>.moveAll(newIndex: Int, predicate: (T) -> Boolean) {
```
Comment (ilya-g) [inline proposals/stdlib/mutablelist-move.md]: I'd also consider as an alternative partitioning the list by predicate and joining the parts again:
```
val reordered = list.partition(predicate).let { it.first + it.second }
```

File: proposals/stdlib/mutablelist-move.md (hunk 5 above)
Comment (ilya-g) [inline proposals/stdlib/mutablelist-move.md]: Imagine you have items with indices 0 and 1 matching the predicate and you want to move them to the end. First you move an item with index 0 to the end and all other items are moved to the beginning. Then you try to move an item with index 1, but it's no more that item that was matched by predicate.

**Comment by thomasnield** [comment]
Started making edits on my end. Will push in the next day or two after I finish.

**Comment by thomasnield** [comment]
Sorry for the delay with this. I'm doing an O'Reilly webcast next week. After that, I'll follow up as my colleagues and I do want this in the stdlib.

**Comment by thomasnield** [comment]
Please refer to this file to see changes. Can you leave any further comments here?
https://github.com/thomasnield/kotlin-collection-extras/blob/master/src/main/kotlin/MutableListMoveExt.kt

**Comment by thomasnield** [comment]
I implemented the proposed changes (I think). Please let me know if I missed anything.

**Comment by ilya-g** [comment]
I still believe these `move*` extensions should not belong to stdlib. 

Given their use cases, the proposed implementations better fit for more scoped UI library operating on `ObservableList` from JavaFX.

Regarding `MutableList.swap` extension, it is considered to be added in https://youtrack.jetbrains.com/issue/KT-2460.

**Comment by thomasnield** [comment]
Fair enough, that sounds valid. I'll track thatbissue and thanks for considering!

**Pull request closed without merging**